cmd/snap-confine/udev-support.c

```diff
@@ -197,12 +204,42 @@ void setup_devices_cgroup(const char *security_tag, struct snappy_udev *udev_s)
 	for (int i = 0; static_devices[i] != NULL; i++)
 		run_snappy_app_dev_add(udev_s, static_devices[i]);
 
-	// add the assigned devices
-	while (udev_s->assigned != NULL) {
-		const char *path = udev_list_entry_get_name(udev_s->assigned);
-		if (path == NULL)
-			die("udev_list_entry_get_name failed");
-		run_snappy_app_dev_add(udev_s, path);
-		udev_s->assigned = udev_list_entry_get_next(udev_s->assigned);
+	// nvidia modules are proprietary and therefore aren't in sysfs and
+	// can't be udev tagged. For now, just add existing nvidia devices to
+	// the cgroup unconditionally (AppArmor will still mediate the access).
+	// We'll want to rethink this if snapd needs to mediate access to other
+	// proprietary devices.
+	struct stat sbuf;
+	char nvpath[16] = { 0 };	// /dev/nvidiaXXX, ctl and -uvm
+	for (unsigned minor = 0; minor < 256; minor++) {
```

Comment:
Can we make this break early when a device is not found?  Would we have gaps?

I feel a bit bad making everybody pay the price for hundreds of syscalls unconditionally when nobody will ever make use even 1% of those calls.

Note that changing the logic will require modifying the "ctl" trick below.

Reply:
I like this idea. I don't think it is possible to have a /dev/nvidia0 and /dev/nvidia3 but not /dev/nvidia2. I also don't think it is possible to have 254 nvidia devices in one machine, but I couldn't come up with a reasonable number. Going sequentially and bailing early sounds fine-- we can always revisit if our assumptions here are wrong.

Reply:
Done